File: rust/everybody-codes/src/bin/2025/y25d11.rs
use everybody_codes::utilities::inputs::get_event_inputs;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input<'a> = &'a str;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_event_inputs(25, 11);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse(input: Input) -> Vec<u64> {
    input.get_numbers().collect()
}

fn solve(mut flock: Vec<u64>, round_limit: Option<usize>) -> (usize, Vec<u64>) {
    let mut changed = true;
    let mut round = 0;
    // phase 1
    while changed {
        round += 1;
        changed = false;
        for i in 0..flock.len() - 1 {
            if flock[i] > flock[i + 1] {
                flock[i] -= 1;
                flock[i + 1] += 1;
                changed = true;
            }
        }
    }
    round -= 1;
    changed = true;
    while changed {
        changed = false;
        for i in 0..flock.len() - 1 {
            if flock[i] < flock[i + 1] {
                flock[i] += 1;
                flock[i + 1] -= 1;
                changed = true;
            }
        }
        round += 1;
        if let Some(limit) = round_limit && round == limit {
            break;
        }
    }
    (round - 1, flock)
}

fn part1(input: Input) -> u64 {
    let flock = parse(input);
    let flock = solve(flock, Some(10)).1;
    flock.into_iter().enumerate().fold(0, |acc, (idx, v)| {
        acc + (idx as u64 + 1) * v
    })
}

fn part2(input: Input) -> usize {
    solve(parse(input), None).0
}

fn part3(input: Input) -> u64 {
    let flock = parse(input);
    let mean = flock.iter().sum::<u64>() / flock.len() as u64;
    flock.into_iter()
        .filter(|&n| n < mean)
        .map(|n| mean - n)
        .sum()
}

#[test]
fn default() {
    let (input1, input2, input3) = get_event_inputs(25, 11);
    assert_eq!(271, part1(&input1));
    assert_eq!(3984738, part2(&input2));
    assert_eq!(130353341887463, part3(&input3));
}

// 1. 271 (8μs)
// 2. 3984738 (145.818ms)
// 3. 130353341887463 (12μs)
// Total: 145.873ms

File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2025/Y25D11.kt
package org.gristle.pdxpuzzles.everybodycodes.y2025

import org.gristle.pdxpuzzles.everybodycodes.utilities.Day
import org.gristle.pdxpuzzles.utilities.parsing.getLongList

object Y25D11 : Day {
    private fun flock(input: String): LongArray {
        val numberList = input.getLongList()
        val iter = numberList.iterator()
        return LongArray(numberList.size) { iter.next() }
    }

    private fun solve(flock: LongArray, roundLimit: Int?): Int {
        var changed = true
        var round = 0
        // phase 1
        while (changed) {
            round++
            changed = false
            for (i in 0 until flock.size - 1) {
                if (flock[i] > flock[i + 1]) {
                    flock[i]--
                    flock[i + 1]++
                    changed = true
                }
            }
        }
        // phase 2
        round -= 1
        changed = true
        while (changed) {
            changed = false
            for (i in 0 until flock.size - 1) {
                if (flock[i] < flock[i + 1]) {
                    flock[i]++
                    flock[i + 1]--
                    changed = true
                }
            }
            round++
            if (round == roundLimit) break
        }
        return round - 1
    }

    override fun part1(input: String): Long {
        val flock = flock(input)
        solve(flock, 10)
        return flock.foldIndexed(0L) { index, acc, v ->
            acc + (index.toLong() + 1) * v
        }
    }
    override fun part2(input: String): Int = solve(flock(input), null)
    override fun part3(input: String): Long {
        val flock = flock(input)
        val mean = flock.sum() / flock.size.toLong()
        return flock
            .filter { n -> n < mean }
            .sumOf { n -> mean - n }
    }
}

fun main() = Day.runDay(Y25D11::class)

//    Quest 1: 271 (0ms)
//    Quest 2: 3984738 (205ms)
//    Quest 3: 130353341887463 (0ms)
//    Total time: 206ms